chacha20/src/backend/neon.rs

```diff
@@ -0,0 +1,620 @@
+    }
+
+    #[inline]
+    #[target_feature(enable = "neon")]
+    unsafe fn store(self, bytes: &mut [u8]) {
+        debug_assert_eq!(bytes.len(), 16);
+
+        // We write the word into a [u32; 4] because vst1q_u32 requires aligned memory.
+        let mut out = [0u32; 4];
+        vst1q_u32(out.as_mut_ptr(), self.0);
+
+        for (word, chunk) in core::array::IntoIter::new(out).zip(bytes.chunks_exact_mut(4)) {
+            chunk.copy_from_slice(&word.to_le_bytes());
+        }
+    }
+
+    #[inline]
+    #[must_use]
+    #[target_feature(enable = "neon")]
+    unsafe fn xor(self, rhs: Self) -> Self {
+        StateWord(veorq_u32(self.0, rhs.0))
+    }
+}
+
+/// A composite type storing the equivalent ChaCha20 word from four consecutive blocks.
+#[derive(Clone, Copy)]
+struct QuadWord(uint32x4_t);
+
+impl QuadWord {
+    #[inline]
+    #[target_feature(enable = "neon")]
+    unsafe fn from_u32(value: u32) -> Self {
+        QuadWord(vdupq_n_u32(value))
+    }
+
+    #[inline]
+    #[must_use]
+    #[target_feature(enable = "neon")]
+    unsafe fn add(self, rhs: Self) -> Self {
+        QuadWord(vaddq_u32(self.0, rhs.0))
+    }
+
+    #[inline]
+    #[must_use]
+    #[target_feature(enable = "neon")]
+    unsafe fn xor(self, rhs: Self) -> Self {
+        QuadWord(veorq_u32(self.0, rhs.0))
+    }
+
+    #[inline]
+    #[must_use]
+    #[target_feature(enable = "neon")]
+    unsafe fn rol<const BY: i32, const REST: i32>(self) -> Self {
+        QuadWord(vsliq_n_u32(vshrq_n_u32(self.0, REST), self.0, BY))
+    }
+
+    #[inline]
+    #[must_use]
+    #[target_feature(enable = "neon")]
+    unsafe fn rol_16(self) -> Self {
+        QuadWord(vreinterpretq_u32_u16(vrev32q_u16(vreinterpretq_u16_u32(
+            self.0,
+        ))))
+    }
+}
+
+/// The state required to process four blocks in parallel.
+///
+/// Unlike the AVX2 backend, where each state word holds a quarter of a single block, here
+/// each quad word holds a specific ChaCha word from all four blocks.
+struct State {
+    x_0: QuadWord,
+    x_1: QuadWord,
+    x_2: QuadWord,
+    x_3: QuadWord,
+    x_4: QuadWord,
+    x_5: QuadWord,
+    x_6: QuadWord,
+    x_7: QuadWord,
+    x_8: QuadWord,
+    x_9: QuadWord,
+    x_10: QuadWord,
+    x_11: QuadWord,
+    x_12: QuadWord,
+    x_13: QuadWord,
+    x_14: QuadWord,
+    x_15: QuadWord,
+}
+
+impl State {
+    /// Stores this state in the given output.
+    #[inline]
+    unsafe fn store(self, output: &mut [u8]) {
+        for (i, (a, b, c, d)) in core::array::IntoIter::new([
+            (self.x_0, self.x_1, self.x_2, self.x_3),
+            (self.x_4, self.x_5, self.x_6, self.x_7),
+            (self.x_8, self.x_9, self.x_10, self.x_11),
+            (self.x_12, self.x_13, self.x_14, self.x_15),
+        ])
+        .enumerate()
+        {
+            let start = 16 * i;
+            let (q_0, q_1, q_2, q_3) = transpose(a, b, c, d);
+
+            for (j, q) in core::array::IntoIter::new([q_0, q_1, q_2, q_3]).enumerate() {
```

Comment:
Is there something I'm missing why `IntoIter` is required here (and also above) and this can't just be...

```suggestion
            for (j, q) in [q_0, q_1, q_2, q_3].iter().enumerate() {
```

They're all copy types, right?

Reply:
Unfortunately no, this results in an iterator of references, for reasons explained in `array::IntoIter`. It was fixed in I think 1.53? But for MSRV 1.51 this approach is necessary.